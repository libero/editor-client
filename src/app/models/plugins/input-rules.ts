import { inputRules, smartQuotes, emDash, ellipsis } from 'prosemirror-inputrules';
import { Plugin } from 'prosemirror-state';

export function buildInputRules(): Plugin {
  const rules = smartQuotes.concat(ellipsis, emDash);
  return inputRules({ rules });
}
