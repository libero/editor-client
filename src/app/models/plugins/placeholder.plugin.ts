import { DecorationSet, Decoration } from 'prosemirror-view';
import { Plugin, EditorState } from 'prosemirror-state';

export function PlaceholderPlugin(placeholderText: string): Plugin {
  return new Plugin({
    props: {
      decorations(state: EditorState) {
        const doc = state.doc;

        if (doc.childCount > 1 || !doc.firstChild.isTextblock || doc.firstChild.content.size > 0) {
          return;
        }

        const placeHolder = document.createElement('div');
        placeHolder.classList.add('placeholder');
        placeHolder.textContent = placeholderText;

        return DecorationSet.create(doc, [Decoration.widget(1, placeHolder)]);
      }
    }
  });
}
