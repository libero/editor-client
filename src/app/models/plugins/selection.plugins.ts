import { Decoration, DecorationSet } from 'prosemirror-view';
import { Plugin } from 'prosemirror-state';

interface DecorationConfig {
  deco: DecorationSet;
}

export const SelectionPlugin = new Plugin({
  state: {
    init(config, instance): DecorationConfig {
      return { deco: DecorationSet.empty };
    },
    apply(transaction, value: DecorationConfig, prevEditorState, editorState): DecorationConfig {
      const sel = transaction.selection;
      if (sel) {
        const decos = [
          Decoration.inline(
            sel.$from.pos,
            sel.$to.pos,
            { class: 'selection-marker' },
            { inclusiveLeft: true, inclusiveRight: true }
          )
        ];
        const deco = DecorationSet.create(editorState.doc, decos);
        return { deco };
      }

      return value;
    }
  },
  props: {
    decorations(state): DecorationSet {
      if (state && this.getState(state)) {
        return this.getState(state).deco;
      }
      return null;
    }
  }
});
