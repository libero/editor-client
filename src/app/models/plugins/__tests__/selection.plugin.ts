import { EditorState, TextSelection } from 'prosemirror-state';
import { Schema } from 'prosemirror-model';

import { SelectionPlugin } from 'app/models/plugins/selection.plugins';
import { EditorView } from 'prosemirror-view';

describe('SelectionPlugin', () => {
  it('should create a decoration', () => {
    const view = createDummyView();
    let state = view.state;
    state = state.apply(state.tr.insertText('Sample text'));
    const changeSel = state.tr.setSelection(TextSelection.create(state.doc, 2, 4));
    state = state.apply(changeSel);

    view.updateState(state);
    expect(view.dom).toMatchSnapshot();
  });

  it('should not create decoration when no selection', () => {
    const view = createDummyView();
    let state = view.state;
    state = state.apply(state.tr.insertText('Sample text'));

    view.updateState(state);
    expect(view.dom).toMatchSnapshot();
  });
});

function createDummyView(): EditorView {
  const state = EditorState.create({
    schema: new Schema({
      nodes: {
        doc: { content: 'inline*' },
        text: { inline: true, group: 'inline' }
      }
    }),
    plugins: [SelectionPlugin]
  });

  const dom = document.createElement('div');
  return new EditorView(dom, {
    dispatchTransaction: jest.fn(),
    state
  });
}
