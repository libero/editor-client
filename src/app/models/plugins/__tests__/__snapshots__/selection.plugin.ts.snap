// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectionPlugin should create a decoration 1`] = `
<div
  class="ProseMirror"
  contenteditable="true"
>
  Sa
  <span
    class="selection-marker"
  >
    mp
  </span>
  le text
</div>
`;

exports[`SelectionPlugin should not create decoration when no selection 1`] = `
<div
  class="ProseMirror"
  contenteditable="true"
>
  Sample text
</div>
`;
